### DISCUSSION TITLE
Offline installation

### DISCUSSION THREAD

**Discussion by Remifa2** [mido/mido#544] [Q&A] [answered]
Hi

I downloaded the mido package mido-1.3.0.tar.gz, but the installation process failed. It seems to need an Internet connection and that's impossible with my gapped PC. The version of Python I'm running is 3.11.1, but I also found a version 3.8.1 installation that came with BespokeSynth. They are in totally different folders.

Here is the info I got :

> py -m pip install "C:\...\mido-1.3.0.tar.gz"
```
Processing c:\...\mido-1.3.0.tar.gz
 Installing build dependencies ... error
  error: subprocess-exited-with-error

  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002220452A310>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
      WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002220451D810>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
      WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002220451D790>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
      WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002220451D750>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
      WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002220452B090>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
      ERROR: Could not find a version that satisfies the requirement setuptools>=61.0.0 (from versions: none)
      ERROR: No matching distribution found for setuptools>=61.0.0
      WARNING: There was an error checking the latest version of pip.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
WARNING: There was an error checking the latest version of pip.
```

I'm new to Python and I don't know how to solve this and ask your kind help.

**Comment by rdoursenaud** [comment, marked as answer]
Hi, sorry to hear that.

I have revamped the packaging method and we now have a few dependencies that we didn't previously.
Unfortunately, there's nothing we can do on our end as this is a `pip` issue.

You'll have to download them all with `pip download` on the Internet connected machine.
You may have to specify `--python-version` & `--platform` for the target machine.
Please see the `pip` documentation: https://pip.pypa.io/en/stable/cli/pip_download/

Don't forget to grab the dependencies for a backend if you want to use ports.
For the default `rtmidi` backend, you'll want to use `pip download mido[ports-rtmidi]` to grab mido and all the required dependencies.
See our documentation for more informations: https://mido.readthedocs.io/en/stable/installing.html

Furthermore, the error message hints at the `setuptools` package being obsolete. You'll want to download and install it first.
As a safety precaution I would also download and install the latest `pip` & `wheel` packages before.

Finally, if you have multiple Python installations, please use the latest version. mido should work with any Python >= 3.7.
Since you seem to be using the Windows python launcher, you can use `py --list` to see all installed versions and specify the version on subsequent invocations:
i.e. for Python 3.11: `py -3.11 -m pip [...]`

Good luck and let me know if you need further help.

Since this is not a bug and the solution may help other users, I'll convert the issue to a discussion.

**Reply by Remifa2** [reply]
I downloaded wheel (https://pypi.org/project/wheel/) not "wheels". Is that ok?

**Reply by rdoursenaud** [reply]
Yes, sorry for the typo. I edited with the correction.

**Reply by Remifa2** [reply]
Thanks for your precious help. I'll report back next week as I'm away for a few days.

**Reply by rdoursenaud** [reply]
You're welcome, hope everything works out!
